### PR TITLE
Update S3 extention codes v2

### PR DESCRIPTION
**Pull request by adam8157**
I did some impovements of s3 after last s3 pull request. This one is to update its code and move development to GPDB source tree.

>  compatible with s3.amazonaws.com and s3-us-east-1.amazonaws.com regions
>  able to cancel query while downloading
>  catch any exceptions thrown by C++ functions
>  set min and max values of threadnum and chunksize
>
>  and will move development to GPDB source tree after this.

v2:
1, separate the large commit (will do small commits later, after this merged, because I was developing on anoter submodule repo which has different hierarchy)
2, report error once an exception is caught
3, expand some comments
4, fix some typos

Cc: @lij55 @yydzero @50wu @hlinnaka @danielgustafsson @eurekaka @xiong-gang @pengzhout 

This pull is the last step of s3ext codes migrating, which blocks following development. Please help me getting this merged quickly. Thanks.

### CODE REVIEW

File: gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp [hunk 1]
```diff
@@ -137,50 +156,74 @@ string S3Reader::getKeyURL(const string key) {
...
+            this->getNextDownloader();
+            if (this->filedownloader) {  // download next file
+                S3INFO("Time to download new file");
+                goto RETRY;
+            }
         }
+        len = buflen;
+    } catch (...) {
+        S3ERROR("Caught an exception, aborting");
+        return false;
```
Comment (yydzero) [inline gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp]: suggest to change messages to 'Caught an exception during transfer data, aborting'. Others LGTM

Comment (adam8157) [inline gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp]: On Tue, Mar 01, 2016 at 08:40:42PM -0800, Yandong Yao wrote:
> suggest to change messages to 'Caught an exception during transfer data,
> aborting'. Others LGTM

Thanks. S3ERROR() will print where it's invoked, that should be fine.

Comment (yydzero) [inline gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp]: Sounds good！

On Wed, Mar 2, 2016 at 12:42 PM, Adam Lee <notifications@github.com> wrote:

> In gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp
> <https://github.com/greenplum-db/gpdb/pull/459#discussion_r54678576>:
>
> >          }
> > +        len = buflen;
> > +    } catch (...) {
> > +        S3ERROR("Caught an exception, aborting");
> > +        return false;
>
> On Tue, Mar 01, 2016 at 08:40:42PM -0800, Yandong Yao wrote: suggest to
> change messages to 'Caught an exception during transfer data, aborting'.
> Others LGTM
> Thanks. S3ERROR() will print where it's invoked, that should be fine.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/459/files#r54678576>.
>



-- 
Best Regards,
Yandong

**Comment by danielgustafsson** [comment]
We don't merge pull requests using the Github interface, please see the wiki on how to merge merges in GPDB for the next PR.

**Comment by hsyuan** [comment]
Can you squash multiple commits into one single commit before pushing it to master? Also please use command line instead of clicking the merge button on the website.

**Pull request merged**